src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs

```diff
@@ -46,71 +49,128 @@ internal class WorkloadUpdateCommand : InstallingWorkloadCommand
                 _workloadInstaller.GetWorkloadInstallationRecordRepository(), _workloadInstaller, _packageSourceLocation, sdkFeatureBand: _sdkFeatureBand);
         }
 
-        public override int Execute()
+        private WorkloadHistoryRecord _WorkloadHistoryRecord
         {
-            if (!string.IsNullOrWhiteSpace(_downloadToCacheOption))
+            get
             {
-                try
+                if (_workloadHistoryState is not null)
                 {
-                    DownloadToOfflineCacheAsync(new DirectoryPath(_downloadToCacheOption), _includePreviews).Wait();
+                    return _workloadHistoryState;
                 }
-                catch (Exception e)
+
+                if (!string.IsNullOrWhiteSpace(_fromHistorySpecified))
                 {
-                    throw new GracefulException(string.Format(LocalizableStrings.WorkloadCacheDownloadFailed, e.Message), e, isUserError: false);
+                    var workloadHistoryRecords = _workloadInstaller.GetWorkloadHistoryRecords().OrderBy(r => r.TimeStarted).ToList();
+                    if (!int.TryParse(_fromHistorySpecified, out int index) || index < 1 || index > workloadHistoryRecords.Count)
```

Comment:
If the index is too large, the current error message text isn't really accurate.